in-progress/8401-proof-timeliness/design.md

```diff
@@ -0,0 +1,268 @@
+    }
+
+    function _prune() internal {
+        uint256 currentSlot = getCurrentSlot();
+        uint256 currentEpoch = getCurrentEpoch();
+        uint256 currentEpochStartSlot = currentEpoch * EPOCH_DURATION;
+
+        if (proofClaim.rewardRecipient == address(0) &&
+            currentSlot >= currentEpochStartSlot + CLAIM_DURATION) {
+            state.pendingTip = state.provenTip;
+            emit PendingChainPruned(state.pendingTip.blockNumber, state.pendingTip.slotNumber);
+        } else if (proofClaim.rewardRecipient != address(0) &&
+                   currentSlot >= currentEpochStartSlot + EPOCH_DURATION) {
+            state.pendingTip = state.provenTip;
+            _slashBond();
+            emit PendingChainPruned(state.pendingTip.blockNumber, state.pendingTip.slotNumber);
+        }
+    }
+
+
+    function proposeBlock(bytes32 _blockHash, bytes32 _archive) external {
+        _prune();
+        // Additional block proposal logic here
+        // ...
+    }
+
+    function submitProofOfEpoch(bytes calldata _epochHeader) external {
+        EpochHeader memory epochHeader = abi.decode(_epochHeader, (EpochHeader));
+        uint256 finalBlockNumber = epochHeader.finalBlockNumber;
+
+        uint256 currentEpoch = getCurrentEpoch();
+
+        // We cannot use chain tip and length because the final block in the epoch may not be in the final slot
+        ProposalLog memory proposalLog = proposalLogs[finalBlockNumber];
+        if (proposalLog.hash == 0) {
+            revert Rollup__NoProposalLog();
+        }
+
+        ProposalLog memory nextProposalLog = proposalLogs[finalBlockNumber + 1];
+        // This serves two checks:
+        // 1. The proof is submitted in the correct epoch
+        // 2. The proof does not truncate the blocks in the epoch it is proving
+        if (nextProposalLog.hash != 0 &&
+            getEpochAt(nextProposalLog.slotNumber) != currentEpoch) {
+            revert Rollup__NotLastBlockInEpoch();
+        }
+
+        bytes32 previousProvenArchive = proposalLogs[state.provenTip.blockNumber].archive;
+        // Proof verification logic here, using the previous proven archive
+        // which ensure's we aren't omitting an epoch
+
+        state.provenTip = ChainTip(finalBlockNumber, proposalLog.slotNumber);
+
+        for (uint256 i = 0; i < epochHeader.fees.length; i++) {
+            Fee memory fee = epochHeader.fees[i];
+            if (fee.recipient != address(0)) {
+                feeJuicePortal.distributeFees(proofClaim.rewardRecipient, fee.recipient, fee.amount);
+            }
+        }
+
+        _returnBond();
+        emit ProofSubmitted(proofClaim.claimedEpoch, finalBlockNumber);
+    }
+
+}
+```
+
+## Change Set
+
+Fill in bullets for each area that will be affected by this change.
+
+- [ ] Cryptography
+- [ ] Noir
+- [ ] Aztec.js
+- [ ] PXE
+- [ ] Aztec.nr
+- [ ] Enshrined L2 Contracts
+- [ ] Private Kernel Circuits
+- [ ] Sequencer
```

Comment:
I know it is saying on L1, but it feels a bit fun to have it on L1, but nothing that is using L1 knows about it 🤷‍♂, e.g., someone needs to know to put up a bond for example, and if we want to use this we would need to also support that part.

Reply:
Agreed. The sequencer changes will be a separate design.